runtime/near-vm-runner/src/cache.rs

```diff
@@ -1,37 +1,144 @@
+        vm_kind: vm_kind as u32,
+    };
+    near_primitives::hash::hash(&key.try_to_vec().unwrap()).try_to_vec().unwrap()
+}
+
+fn cache_error(error: VMError, key: &Vec<u8>, cache: &dyn CompiledContractCache) -> VMError {
+    let record = CacheRecord::Error(error.clone());
+    if cache.put(key.as_slice(), &record.try_to_vec().unwrap()).is_err() {
+        // That's fine, just cannot cache compilation error.
+        println!("Cannot cache an error");
+    }
+    error
+}
+
+fn compile_and_serialize_wasmer(
+    wasm_code: &[u8],
+    config: &VMConfig,
+    key: &Vec<u8>,
+    cache: &dyn CompiledContractCache,
+) -> Result<wasmer_runtime::Module, VMError> {
+    let module = compile_module(wasm_code, config).map_err(|e| cache_error(e, &key, cache))?;
+    let artifact = module
+        .cache()
+        .map_err(|_e| cache_error(VMError::CacheError(SerializationError), &key, cache))?;
+    let code = artifact
+        .serialize()
+        .map_err(|_e| cache_error(VMError::CacheError(SerializationError), &key, cache))?;
+    // If errors comes from serialization we shall not cache it.
+    let serialized = CacheRecord::Code(code)
+        .try_to_vec()
+        .map_err(|_e| VMError::CacheError(SerializationError))?;
+    cache.put(key.as_slice(), &serialized).map_err(|_e| VMError::CacheError(WriteError))?;
+    Ok(module)
+}
+
+fn deserialize_wasmer(serialized: &[u8]) -> Result<wasmer_runtime::Module, VMError> {
+    let record = CacheRecord::try_from_slice(serialized)
+        .map_err(|_e| VMError::CacheError(DeserializationError))?;
+    let code = match record {
+        CacheRecord::Error(err) => return Err(err),
+        CacheRecord::Code(code) => code,
+    };
+    let artifact = Artifact::deserialize(code.as_slice())
+        .map_err(|_e| VMError::CacheError(DeserializationError))?;
+    unsafe {
+        let compiler = compiler_for_backend(Backend::Singlepass).unwrap();
+        load_cache_with(artifact, compiler.as_ref())
+            .map_err(|_e| VMError::CacheError(DeserializationError))
+    }
+}
+
+pub(crate) fn compile_module_cached_wasmer(
+    wasm_code_hash: &[u8],
+    wasm_code: &[u8],
+    config: &VMConfig,
+    cache: Option<&dyn CompiledContractCache>,
+) -> Result<wasmer_runtime::Module, VMError> {
+    /* Consider adding `|| cfg!(feature = "no_cache")` */
+    if cache.is_none() {
+        return compile_module(wasm_code, config);
+    }
+    let key = get_key(wasm_code_hash, wasm_code, VMKind::Wasmer, config);
+    let cache = cache.unwrap();
+    match cache.get(key.as_slice()) {
+        Ok(serialized) => match serialized {
+            Some(serialized) => {
+                match deserialize_wasmer(serialized.as_slice()) {
+                    Ok(module) => Ok(module),
+                    // We trying to be extra careful, and if cannot deserialize code,
+                    // try to serialize again.
+                    Err(_e) => compile_and_serialize_wasmer(wasm_code, config, &key, cache),
```

Comment:
Instead if we can't deserialize, then we should panic. Just throw a CacheError.

Reply:
Well, not sure if it is indeed better. Imagine corrupted DB, we could recover, and no harm is done by this code.

Reply:
For every DB corruption, it's safer to panic. Your node can always resync later.